### ISSUE TITLE
Godot Physics2D misses collisions sometimes when BVH is enabled

### ISSUE THREAD

**Issue by BrianMWest** [godotengine/godot#54855]
### Godot version

3.4.stable

### System information

MacOS 12

### Issue description

I combed through the bugs in https://github.com/godotengine/godot/issues/45334 and I couldn't find anything that seemed to match what I experienced. The closest I could find was https://github.com/godotengine/godot/issues/53997. It's not exactly the same issue, but I'm thinking they might be related because they both seem to involve BVH's internal data becoming stale and collisions either not getting checked or not getting reported.

The player is a `KinematicBody2D` on layer 1. All movement happens with `move_and_collide` with the exception of when I reset the room when a player dies or selects it from the menu (example video is below), in which case all the positions are reset from a signal like so:

```
onready var reset_position := global_position

func _room_reset():
    global_position = reset_position
```

The enemy is an `Area2D` with `monitoring` set to true, `monitorable` set to false, no `layer`, and `mask` set to 1. I tried using both the `on_body_entered` signal and the `overlaps_body` method, although I assume they're heavily linked, so I'm not surprised that neither worked when one didn't. When the player enters the area, it triggers a hit on the enemy, and the enemy flies to the next section of the room. If you reset the room after the enemy has moved to the next section (which is pretty far away), the area no longer detects the player entering. If you reset the room before the enemy flies to the next section, there is no issue. If I fly around right next to the enemy for several seconds, the physics backend seems to eventually figure it out, but it takes 5-10 seconds, and if you reset the room before that, it seems like it starts that timer all over. I've been able to consistently reproduce this bug whenever I follow those steps. When I disable BVH, I can no longer reproduce the bug, so it seems to be something going on with BVH.

Here is a video demonstrating the bug. I am running physics at 120fps, and I tried other values to experiment. I have verified the per-frame movement doesn't even come close to skipping over the `Area2D` in a single frame. The farthest single frame movement is 9 pixels, and the `Area2D` centered on the enemy ship has a circle shape with radius 128. I can consistently reproduce this bug if I reset the room after he flies off screen. If I disable BVH, I can't reproduce it anymore.

https://twitter.com/resinten/status/1458262222367563783

### Steps to reproduce

- create a `KinematicBody2D` (layer = 1) and an `Area2D` (monitorable is False, no layers set, mask = 1)
- have them start somewhat near each other, but not touching
- move both and verify `overlaps_body` returns true on the `Area2D`.
- lerp `Area2D` far away (~2000 pixels)
- reset both back to their starting positions using `global_position` (I use an `onready var reset_position = global_position` to achieve this)
- move them toegether again and collision is no longer detected

**Comment by pouleyKetchoupp** [comment]
Could you please share a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project)? This issue is going to be difficult to investigate otherwise.

**Comment by lawnjelly** [comment]
This is good as it reminds me of the task to add templated checks to the BVH! :grin: Will do after the RIDs are sorted. But yes until then a minimum reproduction project will be useful.

**Comment by BrianMWest** [comment]
I'll see what I can do about getting a minimal reproduction project, but it'll take a bit of time. Right now the project is up to 14k lines of GDScript and 200 scenes, so I'll have to spend cycles to extract just the relevant pieces

**Comment by pouleyKetchoupp** [comment]
@BrianMWest Thank you! For now there's another bug report that might be related (#54898) with a MRP, so I'll start with investigating that one. Once there's a fix you can test your use case, hopefully it will be fixed as well, and if not we might need a MRP from you.

**Comment by pouleyKetchoupp** [comment]
@BrianMWest Please confirm that #54925 helps with your use case if possible. Given the issue I've found it's very likely to be the same one as #54898, but I'd like to be sure.

You can either make a local 3.x build using the patch from that PR, or download a build that corresponds to your platform from the Checks section (once builds are done).

**Comment by ICatRegister** [comment]
> @BrianMWest Please confirm that #54925 helps with your use case if possible

Tried this and it doesn't work for me. Everything is the same, Area2D collisions do not work, but disabling / enabling CollisionShape2D object makes all collisions work immediately

**Comment by BrianMWest** [comment]
> @BrianMWest Please confirm that #54925 helps with your use case if possible. Given the issue I've found it's very likely to be the same one as #54898, but I'd like to be sure.
> 
> You can either make a local 3.x build using the patch from that PR, or download a build that corresponds to your platform from the Checks section (once builds are done).

Unfortunately, this does not fix the bug for me. Working on getting a minimal reproduction project for you now

**Comment by BrianMWest** [comment]
@pouleyKetchoupp Here is a reproduction project. I tried to nail down if there was a consistent distance that caused it to fail, but I couldn't pinpoint it, but it tended to happen if I made the player go just past the enemy, stopped it, and then made the enemy go up to a distance of 2500-3000 before stopping it and resetting the positions. Then subsequent overlaps would fail to be detected. The farther away I let the enemy go, the more likely it would fail when I reset the position.
[bvh-bug-reproduction.zip](https://github.com/godotengine/godot/files/7534793/bvh-bug-reproduction.zip)
